test/default.nix

```diff
@@ -7,49 +7,197 @@
 with pkgs;
 
 let
-  withIfdInputs = builtins.mapAttrs (n: x:
-    if n == "ifdInputs"
-      then pkgs.recurseIntoAttrs (builtins.mapAttrs (_: pkgs.haskell-nix.withInputs) x)
-      else x);
+  # Set recurseForDerivations for both children and grand-children values in
+  # the input association list, but only if the key is "ifdInputs".
+  #
+  # withIfdInputs :: AttrSet -> AttrSet
+  #
+  # The values in the input attribute set must be attribute sets themselves.
+  #
+  # >>> withIfdInputs { ifdInputs = { plan-nix = { a = true; b = "hello"; }; }; cat = 2; }
+  # { ifdInputs = {
+  #     plan-nix = {
+  #       a = true;
+  #       b = "hello";
+  #       recurseForDerivations = true;
+  #     };
+  #     recurseForDerivations = true;
+  #   };
+  #   cat = 2;
+  # }
+  #
+  # >>> withIfdInputs { dog = "hello"; }
+  # { dog = "hello"; }
+  #
+  # >>> withIfdInputs { }
+  # { }
+  withIfdInputs =
+    builtins.mapAttrs
+      (name: val:
+        if name == "ifdInputs"
+        then
+          pkgs.recurseIntoAttrs
+            (builtins.mapAttrs (_: v: pkgs.haskell-nix.withInputs v) val)
+        else val
+      );
+
   util = import ./util.nix { inherit (pkgs.haskell-nix) cabal-install; };
-in pkgs.recurseIntoAttrs {
-  haskellNixRoots = haskell-nix.haskellNixRoots' ifdLevel;
-} // pkgs.lib.optionalAttrs (ifdLevel > 1) (
-  builtins.mapAttrs (_: y: withIfdInputs y) ((if ifdLevel < 3
-    then builtins.mapAttrs (_: d: pkgs.recurseIntoAttrs (pkgs.lib.filterAttrs (n: _: n == "ifdInputs") d))
-    else x: x) {
-  cabal-simple = haskell-nix.callPackage ./cabal-simple { inherit util; };
```

Comment:
Aside from some simple formatting changes, this PR basically just pulls out each part of this code block into separate functions, and adds documentation describing what each does.